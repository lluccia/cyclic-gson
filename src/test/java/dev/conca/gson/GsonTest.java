package dev.conca.gson;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonSerializer;
import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.approvaltests.core.Scrubber;
import org.approvaltests.scrubbers.RegExScrubber;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class GsonTest {

    private Gson gson;

    @BeforeEach
    void setUp() {
        gson = new Gson();
    }

    @Test
    void canSerializeEmptyObject() {
        CyclicObject rootObject = new CyclicObject();
        assertThat(gson.toJson(rootObject)).isEqualTo("{}");
    }

    @Test
    void canSerializeSimpleFields() {
        CyclicObject rootObject = new CyclicObject();
        rootObject.setField1("value1");
        assertThat(gson.toJson(rootObject)).isEqualTo(
                "{\"field1\":\"value1\"}");
    }

    @Test
    void canGetUniqueInstanceIdentifiers() {
        CyclicObject object1 = new CyclicObject();
        object1.setField1("value1");

        CyclicObject object2 = object1;

        CyclicObject object3 = new CyclicObject();
        object3.setField1("value1");

        assertThat(System.identityHashCode(object1))
                .isEqualTo(System.identityHashCode(object2))
                .isNotEqualTo(System.identityHashCode(object3));
    }

    @Test
    void shouldSerializeIdentifiersForReferencedObjects() {
        JsonSerializer<CyclicObject> serializer = new CyclicObjectSerializer();

        Gson customGson = new GsonBuilder()
                .setPrettyPrinting()
                .registerTypeAdapter(CyclicObject.class, serializer)
                .create();

        CyclicObject object1 = new CyclicObject();
        object1.setField1("object1");

        CyclicObject object2 = new CyclicObject();
        object2.setField1("object2");

        object1.setCyclicObject(object2);
        object2.setCyclicObject(object1);

        Scrubber objectIdScrubber = new RegExScrubber("#\\d+", n -> "#[id" + n + "]");
        Approvals.verify(customGson.toJson(object1), new Options(objectIdScrubber));
    }

}