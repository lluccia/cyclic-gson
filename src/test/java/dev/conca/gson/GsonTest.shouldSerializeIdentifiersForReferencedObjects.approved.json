{
  "dev.conca.gson.CyclicObject#[id1]": {
    "field1": "object2",
    "cyclicObject": {
      "$ref": "/dev.conca.gson.CyclicObject#[id2]"
    }
  },
  "dev.conca.gson.CyclicObject#[id2]": {
    "field1": "object1",
    "cyclicObject": {
      "$ref": "/dev.conca.gson.CyclicObject#[id1]"
    }
  }
}
